### ISSUE TITLE
pre-commit 2.0

### ISSUE THREAD

**Issue by asottile** [pre-commit/pre-commit#1260]
This will be the first major breaking release of `pre-commit`

One final feature release (likely v1.21) will be released before 2.0

These are the planned removals / changes in 2.0:

### dropping old python versions

*note*: despite the `pre-commit` binary dropping older python versions, it will still be able to run hooks written in those versions as long as `virtualenv` retains support for those

the plan is to require **python3.6+**, this will allow a couple of things:
- removal of python2-specific code
- type annotations (including variable annotations)

### removal of the `pcre` language

- `pcre` has not-the-best cross platform story (broken on windows for particular regexes, requires specific package to be installed via brew on macos)
- `pygrep` in most cases is a drop-in replacement for PCRE
- `pygrep` was introduced on [2017-10-03](https://github.com/pre-commit/pre-commit/blob/master/CHANGELOG.md#120---2017-10-03) (2 years + 2 months)
- `pcre` has been deprecated since [2018-01-13](https://github.com/pre-commit/pre-commit/blob/master/CHANGELOG.md#150---2018-01-13) (1 year + 11 months)

### removal of `pre-commit autoupdate --tags-only`

- this has been the default behaviour and has produced a warning since [2017-05-16](https://github.com/pre-commit/pre-commit/blob/master/CHANGELOG.md#0140---2017-05-16) (2 years + 7 months)

___

### :x: .pre-commit-config.yaml is always a top-level map

a warning has not been produced for this, it is [still marked as TODO](https://github.com/pre-commit/pre-commit/blob/2500767f1bbd133d2be3b1c6f29ad91c49f39b7c/pre_commit/clientlib.py#L289-L293) and as such cannot break yet

### :x: `sha` -> `rev`

- `rev` has been available since [2018-03-03](https://github.com/pre-commit/pre-commit/blob/master/CHANGELOG.md#170---2018-03-03) (1 year, 9 months)
- started producing a deprecation warning since [2019-06-06](https://github.com/pre-commit/pre-commit/blob/master/CHANGELOG.md#1170---2019-06-06) (6 months)

**Comment by asottile** [comment]
If you'd like continued support for older pythons, please reach out to me privately so we can discuss that -- you can find my email address through git commits in this repository or DM me on [twitter](https://twitter.com/codewithanthony)

**Comment by ssbarnea** [comment]
I would prefer the python dropping to be delayed, but regarding hiding config file from root, I wonder if there is any work done standardize it to something like a `.config`. That feature would be really nice but it makes sense only if other tool maintainers agree on a migration plan (towards a documented location).

**Comment by asottile** [comment]
please reach out privately thanks -- your other comment is a feature request which is completely off topic and not something being pursued by this task

**Comment by asottile** [comment]
via 0cc199d351ab126abd874a4220b4f6c11362ee71

**Comment by asottile** [comment]
fwiw, this is what you'll see for the various breaking changes:

### using python2.x

```console
$ pre-commit --help
Traceback (most recent call last):
  File "/tmp/x/venv/bin/pre-commit", line 7, in <module>
    from pre_commit.main import main
  File "/tmp/x/venv/local/lib/python2.7/site-packages/pre_commit/main.py", line 44
    def _add_color_option(parser: argparse.ArgumentParser) -> None:
                                ^
SyntaxError: invalid syntax
```

(make sure to upgrade to `pip>=9` which will avoid installing incompatible versions)

### using python3.5

```console
$ pre-commit --help
Traceback (most recent call last):
  File "/tmp/x/venv/bin/pre-commit", line 7, in <module>
    from pre_commit.main import main
  File "/tmp/x/venv/lib/python3.5/site-packages/pre_commit/main.py", line 168
    version=f'%(prog)s {C.VERSION}',
                                  ^
SyntaxError: invalid syntax
```

(make sure to upgrade to `pip>=9` which will avoid installing incompatible versions)

### pcre

```console
$ pre-commit  run foo --all-files
An error has occurred: InvalidConfigError: 
==> File .pre-commit-config.yaml
==> At Config()
==> At key: repos
==> At Repository(repo='local')
==> At key: hooks
==> At Hook(id='foo')
==> At key: language
=====> Expected one of conda, docker, docker_image, fail, golang, node, pygrep, python, python_venv, ruby, rust, script, swift, system but got: 'pcre'
Check the log at /home/asottile/.cache/pre-commit/pre-commit.log
```

~usually you can replace `pcre` hooks directly with `pygrep` hooks

### `--tags-only`

```console
$ pre-commit autoupdate --tags-only
usage: pre-commit [-h] [-V]
                  {autoupdate,clean,hook-impl,gc,init-templatedir,install,install-hooks,migrate-config,run,sample-config,try-repo,uninstall,help}
                  ...
pre-commit: error: unrecognized arguments: --tags-only
```

(don't use the argument, it does nothing now)

**Comment by obestwalter** [comment]
Thanks for releasing a new shiny version of this great tool :champagne:

With Python 3.6.0 we ran into: 

```text
pre-commit version: 2.1.0
sys.version:
    3.6.0 (default, Jan  3 2018, 14:56:17)
    [GCC 4.6.4]
sys.executable: /home/[...]/.tox/lint/bin/python3.6
os.name: posix
sys.platform: linux

### error information

An unexpected error has occurred: AttributeError: type object 'Hook' has no attribute 'create'

Traceback (most recent call last):
  File "/home/[...]/.tox/lint/lib/python3.6/site-packages/pre_commit/error_handler.py", line 54, in error_handler
    yield
  File "/home/[...]/.tox/lint/lib/python3.6/site-packages/pre_commit/main.py", line 371, in main
    return run(args.config, store, args)
  File "/home/[...]/.tox/lint/lib/python3.6/site-packages/pre_commit/commands/run.py", line 326, in run
    for hook in all_hooks(config, store)
  File "/home/[...]/.tox/lint/lib/python3.6/site-packages/pre_commit/repository.py", line 206, in all_hooks
    for repo in root_config['repos']
  File "/home/[...]/.tox/lint/lib/python3.6/site-packages/pre_commit/repository.py", line 207, in <genexpr>
    for hook in _repository_hooks(repo, store, root_config)
  File "/home/[...]/.tox/lint/lib/python3.6/site-packages/pre_commit/repository.py", line 182, in _repository_hooks
    return _cloned_repository_hooks(repo_config, store, root_config)
  File "/home/[...]/.tox/lint/lib/python3.6/site-packages/pre_commit/repository.py", line 170, in _cloned_repository_hooks
    for hook in hook_dcts
  File "/home/[...]/.tox/lint/lib/python3.6/site-packages/pre_commit/repository.py", line 170, in <genexpr>
    for hook in hook_dcts
AttributeError: type object 'Hook' has no attribute 'create'
```

With the newest patch version of Python 3.6 the error does not occur anymore. I managed to dig up a 3.6.4 from somewhere and that also worked. Don't know if it would make sense to bump the minimum requirements to 3.6.4 or something ...

**Comment by asottile** [comment]
oof, I'll release a 2.1.1 with a fix for this and then bump the `python_requires` for the 2.2 release

I'll also follow up and add this check to [flake8-typing-imports](https://github.com/asottile/flake8-typing-imports) -- didn't know that `NamedTuple` didn't support methods until 3.6.1

**Comment by asottile** [comment]
this has been released in [v2.1.1](https://github.com/pre-commit/pre-commit/releases/tag/v2.1.1) -- thanks for the report! :tada:

**Comment by asottile** [comment]
and here's the patch to `flake8-typing-imports` to check for this: https://github.com/asottile/flake8-typing-imports/pull/14
